rfcs/20230517-forward-compatibility-testing.md

```diff
@@ -0,0 +1,102 @@
+# StableHLO Forward Compatibility Testing
+
+## Background
+
+StableHLO needs to verify both forward and backward compatibility via unit
+tests. Currently we check backward compatibility by versioning
+`stablehlo_legalize_to_vhlo.mlir`. This document makes the following proposals:
+
+* **Proposal 1:** Maintain bytecode at HEAD to use for forward compatibility
+testing.
+* **Proposal 2:** Add CI which builds at tagged release commits to check
+forward compatibility.
+* **Proposal 3:** Run CI on all supported minor releases of StableHLO
+(`v0.9.0`, `v0.10.0`, etc.).
+
+## Requirements
+
+Forward compatibility testing has the following requirements, to ensure that it
+is both useful and practical:
+
+* **R1.** Tests must detect changes in StableHLO that break forward compatibility.
+* **R2.** Tests must detect LLVM revision bumps that break forward compatibility.
+* **R3.** Test job must scale with releases and execute in a reasonable amount
+of time.
+
+## Preferred Design: Serialize at HEAD, Deserialize at Target Release
+
+_Note: Section only discusses how to test a generic release `0.X.0`. Discussion
+on what values of `X` to test for are discussed in the later section "What
+releases should be tested?"_
+
+### Serialize and Commit Test Files at HEAD
+
+Prior to tagging each StableHLO release, which occurs ~2x per week, generate
+versioned forward compatibility tests using the
+`stablehlo_legalize_to_vhlo.0_X_0.mlir` test file as follows:
+
+```bash
+stablehlo-translate --serialize --target=0.X.0
+  stablehlo_legalize_to_vhlo.0_X_0.mlir >
+    stablehlo_legalize_to_vhlo.0_X_0.mlir.bc.HEAD
+```
+
+The output will be a bytecode file that targets `0.X.0` and should be runnable
+at release `0.X.0`. Once these files are generated and checked in, we can add
+the following RUN line to the source `.mlir` file to test serialization,
+deserialization, and FileCheck comparisons:
+
+```bash
+// RUN: stablehlo-translate --deserialize %s.bc.HEAD |
+  stablehlo-translate --serialize --target=0.X.0 |
+    stablehlo-opt --mlir-print-op-generic | FileCheck %s
+```
+
+**Proposal 1:** Maintain bytecode at HEAD to use for forward compatibility
+testing.
+
+### Add Forward Compatibility CI
+
+A CI job which blocks pull request merges to main can be added to ensure
+forward compatibility of PRs. To test version `0.X.0` the CI job does the
+following:
+
+* Checkout `openxla/stablehlo` at tag `v0.X.0`.
+* Checkout file at HEAD `stablehlo_legalize_to_vhlo.0_X_0.mlir.bc.HEAD`
```

Comment:
We could. I think this would require a single CI job which builds at HEAD, serializes, then checks out older revisions of StableHLO/LLVM and rebuilds sequentially. This can't be parallelized as easily as the current approach (see prototype).

Alternatively this would require each CI job to rebuild at HEAD then check out an older tag and rebuild StableHLO/LLVM at those revisions. Slightly higher parallelism, but more things being built. 

Either way, this is being reworked into the alternative design, so lets explore the proposed design before optimizing this approach.